workflows/mapping/Snakefile

```diff
@@ -0,0 +1,266 @@
+    return os.path.join(HERE, WRAPPERS, tool)
+
+
+def params_for(rule, key):
+    d = config.get('rules', {}).get(rule, {}).get('params', {}).get(key, '')
+    return d
+
+
+def threads_for(rule):
+    return config.get('rules', {}).get(rule, {}).get('threads', {})
+
+
+def fastq_screen_config_inputs():
+    """
+    Returns indexes and a table (as string) that can be written directly to
+    a config file.
+
+    Use indexes as the input to a rule; use the table to build a config file.
+    """
+    indexes = config['rules']['fastq_screen']['indexes']
+    aligner = config['rules']['fastq_screen']['aligner']
+    inputs = []
+    for label, prefix in indexes.items():
+        prefix = os.path.join(config['data_dir'], prefix)
+        inputs.extend(expand(prefix + '.{n}.bt2', n=[1, 2, 3, 4]))
+    return inputs
+
+
+sampletable = pandas.read_table(config['sampletable'], index_col=0)
+samples = sampletable.index.tolist()
+
+include: '../references/references.snakefile'
+
+
+# ----------------------------------------------------------------------------
+# Aligner configuration.
+#
+ALIGNER_FASTQ_PATTERN = '{sample_dir}/{sample}/{sample}_R1.cutadapt.fastq.gz'
+ALIGNER_BAM_OUTPUT_PATTERN = '{{sample_dir}}/{{sample}}/{{sample}}.cutadapt.{0}.bam'
+ALIGNER_LOG_OUTPUT_PATTERN = '{{sample_dir}}/{{sample}}/{{sample}}.cutadapt.{0}.bam.log'
+ALIGNER = config['rules']['align']['aligner']
+ALIGNER_PREFIX = config['rules']['align']['prefix']
+ALIGNER_TAG= {'bowtie2': 'bt2',  'hisat2': 'ht2'}[ALIGNER]
+
+if ALIGNER == 'hisat2':
+    ALIGNER_INDEX = expand('{data_dir}/{prefix}.{n}.ht2',
+                           data_dir=config['data_dir'], prefix=ALIGNER_PREFIX,
+                           n=range(1, 9))
+
+if ALIGNER == 'bowtie2':
+    ALIGNER_INDEX = expand('{data_dir}/{prefix}.{n}.bt2',
+                           data_dir=config['data_dir'], prefix=ALIGNER_PREFIX,
+                           n=range(1, 5))
+
+# ----------------------------------------------------------------------------
+# Generate final targets
+#
+patterns = [
+    '{sample_dir}/{sample}/{sample}_R1.fastqc.html',
+    '{sample_dir}/{sample}/{sample}_R1.fastq_screen.txt',
+    '{sample_dir}/{sample}/{sample}_R1.fastq_screen.png',
+    '{sample_dir}/{sample}/{sample}_R1.cutadapt.fastqc.html',
+    '{sample_dir}/{sample}/{sample}_R1.cutadapt.fastq.gz',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.bam',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.bam.observed_complexity',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.bam.estimated_complexity',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.bam.estimated_coverage',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.sort.dedup.bam',
+    '{sample_dir}/{sample}/{sample}.cutadapt.{aligner_tag}.unique.sort.dedup.metrics',
+    'multiqc_report.html',
+    'fastq_screen_config.txt',
+    fastq_screen_config_inputs(),
+    ALIGNER_INDEX,
+]
```

Comment:
Not that important here, but what if someone does not want to run cutadapt? Just a thinking place holder.